### PR TITLE
[Lang] MatrixNdarray refactor part9: Add scalarization for AllocaStmt

### PR DESCRIPTION
**Pull request by jim19930609**
Related issue = https://github.com/taichi-dev/taichi/issues/5873, https://github.com/taichi-dev/taichi/issues/5819

This PR is working "Part ④" in https://github.com/taichi-dev/taichi/issues/5873.

[AllocaStmt scalarization]
```
Before:
  TensorType<4 x i32>* addr = AllocaStmt(TensorType<4 x i32>)

After:
  i32 addr0 = AllocaStmt(i32)
  i32 addr1 = AllocaStmt(i32)
  i32 addr2 = AllocaStmt(i32)
  i32 addr3 = AllocaStmt(i32)

  scalarized_local_tensor_map_[addr] = {addr0, addr1, addr2, addr3}
```

[Load AllocaStmt]
```
Before:
  TensorType<4 x i32> val = LoadStmt(TensorType<4 x i32>* alloca_src)

After:
  i32 val0 = LoadStmt(scalarized_local_tensor_map_[stmt][0])
  i32 val1 = LoadStmt(scalarized_local_tensor_map_[stmt][1])
  i32 val2 = LoadStmt(scalarized_local_tensor_map_[stmt][2])
  i32 val3 = LoadStmt(scalarized_local_tensor_map_[stmt][3])

  tmp = MatrixInitStmt(val0, val1, val2, val3)
  stmt->replace_all_usages_with(tmp)
```

[Store to AllocaStmt]
```
Before:
  StoreStmt(TensorType<4 x i32>* alloca_dest_stmt, TensorType<4 x i32> val)

After:
  StoreStmt(i32* scalarized_local_tensor_map_[stmt][0], 
            i32 val->cast<MatrixInitStmt>()->val[0]) 
  StoreStmt(i32* scalarized_local_tensor_map_[stmt][1], 
            i32 val->cast<MatrixInitStmt>()->val[1]) 
  StoreStmt(i32* scalarized_local_tensor_map_[stmt][2], 
            i32 val->cast<MatrixInitStmt>()->val[2]) 
  StoreStmt(i32* scalarized_local_tensor_map_[stmt][3], 
            i32 val->cast<MatrixInitStmt>()->val[3])
```

### CODE REVIEW

**Comment by jim19930609** [comment]
Gonna add python tests in a following PR~

**Comment by AD1024** [comment]
Thanks for implementing this!
Are we leveraging the CFG pass to eleminate that MatrixInitStmt holding pointers to scalarized values?

**Comment by jim19930609** [comment]
> Thanks for implementing this! LGTM. Are we leveraging the CFG pass to eleminate that MatrixInitStmt holding pointers to scalarized values?

I dont think MatrixInitStmt is holding any pointer-typed values? Are you referring to the redundant MatrixInitStmt inserted during scalarization, those are remove right after scalarization is done (There's will be MatrixInitCleanUp pass).

**Review by strongoier — COMMENTED** [review]
Hmm.. I think having a case in the `LowerMatrixPtr` pass is much easier? You don't have to modify the `Scalarize` pass at all.

**Comment by jim19930609** [comment]
> Hmm.. I think having a case in the `LowerMatrixPtr` pass is much easier? You don't have to modify the `Scalarize` pass at all.

But it is trying to scalarize the `AllocaStmt(TensorType)` into `AllocaStmt(scalar0), ...` thus well fits the scalarized pass? 

The other thing with `LowerMatrixPtr` is that it doesn't rely on whether "scalarized" is turned-on or not, but scalarization for `AllocaStmt` should only be enabled when we decide to do so, thus I feel it's better to decouple these two passes?

**Comment by strongoier** [comment]
> But it is trying to scalarize the `AllocaStmt(TensorType)` into `AllocaStmt(scalar0), ...` thus well fits the scalarization pass?

IIUC `Scalarization` aims at faking a tensor operation with a bunch of scalar operations, and `LowerMatrixPtr` aims at eliminating intermediate matrix ptrs. I agree that `AllocaStmt(TensorType) -> AllocaStmt(scalar0), ...` should be put in `Scalarization`, but I think `PtrOffsetStmt(AllocaStmt)` should be better handled in `LowerMatrixPtr`. Thus theoretically we need an intermediate representation like `MatrixOfPtrStmt`.

However, considering that `AllocaStmt(TensorType) -> AllocaStmt(scalar0), ...` may not be needed in the end, and we haven't decided which parts are mandatory or optional, I think we don't need to be so strict with pass separation for now. The only suggestion here may be to simplify the implementation a bit - `visit(PtrOffsetStmt)` should be able to cover the changes in `visit(LoadStmt) / visit(StoreStmt) / ...`?

**Comment by jim19930609** [comment]
> > But it is trying to scalarize the `AllocaStmt(TensorType)` into `AllocaStmt(scalar0), ...` thus well fits the scalarization pass?
> 
> IIUC `Scalarization` aims at faking a tensor operation with a bunch of scalar operations, and `LowerMatrixPtr` aims at eliminating intermediate matrix ptrs. I agree that `AllocaStmt(TensorType) -> AllocaStmt(scalar0), ...` should be put in `Scalarization`, but I think `PtrOffsetStmt(AllocaStmt)` should be better handled in `LowerMatrixPtr`. Thus theoretically we need an intermediate representation like `MatrixOfPtrStmt`.
> 
> However, considering that `AllocaStmt(TensorType) -> AllocaStmt(scalar0), ...` may not be needed in the end, and we haven't decided which parts are mandatory or optional, I think we don't need to be so strict with pass separation for now. The only suggestion here may be to simplify the implementation a bit - it seems that `visit(PtrOffsetStmt)` already covers the changes in `visit(LoadStmt) / visit(StoreStmt) / ...`?

Simplified `irpass::scalarize()` but still kept scalarization for `AllocaStmt` and `PtrOffsetStmt` as part of `irpass::scalarize()`. The barrier to split "scalarization for AllocaStmt" and "optimization for PtrOffsetStmt(AllocaStmt)" lies in the fact that we need this `scalarized_local_tensor_map_` to record the mapping from original `AllocaStmt` to the scalarized ones, and transferring this `scalarized_local_tensor_map_` across passes will be implementation-wisely ugly.

File: tests/cpp/transforms/scalarize_test.cpp [hunk 1]
```diff
@@ -90,49 +141,89 @@ void test_load_scalarize() {
...
+  EXPECT_EQ(block->statements[7]->is<LocalStoreStmt>(), true);
+  EXPECT_EQ(block->statements[8]->is<LocalStoreStmt>(), true);
+  EXPECT_EQ(block->statements[9]->is<LocalStoreStmt>(), true);
+  EXPECT_EQ(block->statements[10]->is<LocalStoreStmt>(), true);
+}
 
-  EXPECT_EQ(block->statements[11]->is<ConstStmt>(), true);
-  EXPECT_EQ(block->statements[12]->is<MatrixPtrStmt>(), true);
-  EXPECT_EQ(block->statements[13]->is<T>(), true);
+TEST(Scalarize, ScalarizeLoadAlloca) {
```
Comment (strongoier) [inline tests/cpp/transforms/scalarize_test.cpp:185]: I didn't quite get the naming - we have `ScalarizeGlobalStore` and `ScalarizeLocalStore` but not `ScalarizeGlobalLoad` and `ScalarizeLocalLoad`?

**Review by strongoier — COMMENTED** [review]
Great!!

File: taichi/transforms/scalarize.cpp [hunk 2]
```diff
@@ -158,9 +191,11 @@ class Scalarize : public IRVisitor {
 
       std::vector<Stmt *> matrix_init_values;
       int num_elements = operand_tensor_type->get_num_elements();
+      auto primitive_type = operand_tensor_type->get_element_type();
       for (size_t i = 0; i < num_elements; i++) {
         auto unary_stmt = std::make_unique<UnaryOpStmt>(
             stmt->op_type, operand_matrix_init_stmt->values[i]);
+        unary_stmt->ret_type = primitive_type;
```
[the remark this answers is not in the document]
Comment (jim19930609) [reply taichi/transforms/scalarize.cpp:198]: Thanks! Didn't realize that~

File: tests/cpp/transforms/scalarize_test.cpp (hunk 1 above)
Comment (jim19930609) [reply tests/cpp/transforms/scalarize_test.cpp:185]: Oh I forgot to test `LocalLoad(AllocaStmt)`. I'll have that fixed. 

In general, `ScalarizedLocalXXX` tests with pointers from `AllocaStmt` and `ScalarizedGlobalXXX` tests with pointers from `ArgLoadStmt`

File: taichi/transforms/scalarize.cpp [hunk 3]
```diff
@@ -244,62 +260,132 @@ class Scalarize : public IRVisitor {
...
-  }
+  // { original_alloca_stmt : [scalarized_alloca_stmt0, ...] }
+  std::unordered_map<Stmt *, std::vector<Stmt *>> scalarized_local_tensor_map_;
 
-  void visit(GlobalStoreStmt *stmt) override {
-    scalarize_store_stmt<GlobalStoreStmt>(stmt);
-  }
+  ScalarizePointers(IRNode *node) {
+    allow_undefined_visitor = true;
+    invoke_default_visitor = false;
```
Comment (strongoier) [inline taichi/transforms/scalarize.cpp:292]:
```suggestion
```
IIRC these are default settings of `BasicStmtVisitor`.

**Review by strongoier — APPROVED** [review]

**Pull request merged**